### ISSUE TITLE
[Discussion] Service Bus - Track 2 - Rule operations

### ISSUE THREAD

**Issue by JoshLove-msft** [Azure/azure-sdk-for-net#10842]
_The intent of this issue is to solicit feedback from the community in advance of the GA of the next major version of the Service Bus library (which will follow several Preview releases).
As a reference, the .NET Azure SDK guidelines can be found here: https://azure.github.io/azure-sdk/dotnet_introduction.html
Terminology:
Terminology:
Track 0 refers to the [WindowsAzure.ServiceBus library](https://www.nuget.org/packages/WindowsAzure.ServiceBus/)
Track 1 refers to the [Microsoft.Azure.ServiceBus library](https://www.nuget.org/packages/Microsoft.Azure.ServiceBus/)
Track 2 refers to the new library, Azure.Messaging.ServiceBus, we are working on releasing._

Rule management operations can be performed for subscriptions. In [version 4](https://www.nuget.org/packages/Microsoft.Azure.ServiceBus/) of the Service Bus library, the rule operations live on the [SubscriptionClient](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Microsoft.Azure.ServiceBus/src/SubscriptionClient.cs). The available rule operations are [AddRuleAsync](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Microsoft.Azure.ServiceBus/src/SubscriptionClient.cs#L464), [RemoveRuleAsync](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Microsoft.Azure.ServiceBus/src/SubscriptionClient.cs#L523), and [GetRulesAsync](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Microsoft.Azure.ServiceBus/src/SubscriptionClient.cs#L564). These management operations differ conceptually from the data operations that you would do with a subscription, namely receiving and settling messages. Hence, it makes sense to separate out these management concerns from the data concerns of receiving/settling. This also allows the ServiceBusReceiver type, which is the type that will be used for receiving from queues/subscriptions, to present a simpler interface. Finally, since we will have one type for receiving from both queues and subscriptions, we can avoid adding validation to prevent users from doing rule operations on a queue, since in order to do any rule management operations, a user will need to supply a topic and subscription name.

This raises the question - why even have these management operations in our AMQP clients at all? Why not just rely on the HTTP management client which supports these operations? Well, it is possible to bundle rule operations together with other data operations in transactions. Our hypothesis is that this is not a common scenario though. It will still be possible to do this bundling even with the change to have the rule management live off of the client, rather than directly on the receiver.

```c#
var client = new ServiceBusClient("connectionString");
var ruleMgr = client.GetRuleManager("topic", "subscription");
var receiver = client.GetReceiver("topic", "subscription", receiverOptions);
using (var ts = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled))
{
    var message = await receiver.ReceiveAsync();
    await receiver.CompleteAsync(message.LockToken);
    await ruleMgr.AddRuleAsync("ruleName", filter);
}
```

**Comment by JoshLove-msft** [comment]
Adding some community members who have recently submitted issues/PRs for Service Bus 😃

/cc @nemakam @SeanFeldman @Einarsson @brandondahler @broloaugh @ankrause @GrimGadget @ShaneCourtrille @Pucmaster @Bnjmn83

**Comment by SeanFeldman** [comment]
Initially, during Track 1 there was a very strong resistance to add the HTTP management client to mimic what Track 0 was offering. Very quickly (and after a long debate with the community) it was very obvious that Microsoft was trying to push something that the real world either didn't want or was not ready for. Long story short, the management client was added back. The AMQP option to manage rules was added much earlier, assuming that those operations are very common at runtime and therefore should be supported. Which goes against the your current hypothesis:

> Well, it is possible to bundle rule operations together with other data operations in transactions. Our hypothesis is that this is not a common scenario though.

First, let's make sure we don't overload the context here by using the word 'transactions'. Perhaps operation, but not transactions. The word 'transaction' implies a feature supported by Azure Service Bus.

Second, this is not as uncommon as you're assuming. As a matter of fact, it is quite common. Rules _are_ added and modified at runtime more than you can imagine. Most of my customers prefer to have those rules configured and modified from the code that executes data operations rather than script or execute it separately.

> It will still be possible to do this bundling even with the change to have the rule management live off of the client, rather than directly on the receiver.

It is what many are doing today. The substantial downside of the HTTP management client as oppose to an AMQP one is speed. HTTP management client is extremely slow.

With that, you should consider getting the feedback from the PG which can elaborate on additional insights of HTTP vs AMQP operations. 

A personal note - having similar (or identical) operations in two locations is confusing. I would be happy to see those going through one channel and one channel only. In general, users prefer to use one management client that has rather than various clients for different types of entities.

**Comment by JoshLove-msft** [comment]
> First, let's make sure we don't overload the context here by using the word 'transactions'. Perhaps operation, but not transactions. The word 'transaction' implies a feature supported by Azure Service Bus.

Yes, I'm specifically referring to the transaction support that Service Bus offers, wherein operations can be grouped together into an atomic unit. The concern with abandoning the rule management from the AMQP layer was that users wouldn't be able to do transactions involving rule management. With the proposal here, users can still do transactions involving rule management, but the rule management operations would live on a separate type than the other data operations.

We haven't gotten to the point of migrating the HTTP management client yet, but the feedback of not wanting to have the same operation available from two locations is helpful.

>With that, you should consider getting the feedback from the PG which can elaborate on additional insights of HTTP vs AMQP operations.

Can you clarify what you mean by "PG"?

**Comment by SeanFeldman** [comment]
PG = Product Group

**Comment by SeanFeldman** [comment]
I've never saw the need to mix message and management operations in a transaction. I also don't know if that's even supported, given that a transaction operates on messages and not entities.

**Comment by JoshLove-msft** [comment]
> I've never saw the need to mix message and management operations in a transaction. I also don't know if that's even supported, given that a transaction operates on messages and not entities.

This was highlighted as a scenario that is supported in the Track 1 libraries - having a transaction that includes both rule operation and message operations. In this proposal, we are still going to support this, but it would involve two types interacting with the service instead of just one.

/cc @nemakam

**Comment by nemakam** [comment]
@SeanFeldman , yeah, there are customers who use update/add/delete rules within a transaction and that does become an important scenario to consider. That's the sole reason for allowing rule management through the regular AMQP path along with the HTTP path.

**Comment by nemakam** [comment]
As for transactions on AMQP, all transactional operations need to be on the same connection. As long as that is satisfied, I'm okay with moving them out to a separate client.

**Comment by SeanFeldman** [comment]
Somewhat related to the conversation - the option of mixing messaging and management operations is not documented anywhere as far as I know. It would be good to document it if that feature survives past track 1. Raised it as a broker issue for now [here](https://github.com/Azure/azure-service-bus/issues/351).

**Comment by JoshLove-msft** [comment]
Thanks for the participation everyone!

Closing this issue as it has served its purpose. We have moved the new [Azure.Messaging.ServiceBus](https://www.nuget.org/packages/Azure.Messaging.ServiceBus/) package out of previews and into a stable release (version 7.0.0).

It turned out that including both messaging and rule management operations as part of a single transaction is not supported. Being that this scenario is not supported, we removed the AMQP rule manager client. Rule management can be performed using the operations available on the `ServiceBusAdministrationClient` which works over HTTP.

See the [migration guide from Microsoft.Azure.ServiceBus to Azure.Messaging.ServiceBus](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Azure.Messaging.ServiceBus/MigrationGuide.md) for more details
